### ISSUE TITLE
Problems with std::map in a Debug build using recent versions of GCC

### ISSUE THREAD

**Issue by hahnjo** [root-project/root#10291]
Enrico originally reported that the following macro
```c++
#include <map>
#include <string>

void repro_modules_crash() {
  std::map<std::string, bool> m{{"foo", true}, {"bar", false}};
}
```
when executed as `root -l repro_modules_crash.C` gives the following errors with a `Debug` build on Arch Linux:
```
In module 'std' imported from input_line_1:1:
/usr/lib/gcc/x86_64-pc-linux-gnu/11.2.0/../../../../include/c++/11.2.0/bits/stl_tree.h:256:12: error: instantiation of 'std::_Rb_tree_iterator<std::pair<const std::__cxx11::basic_string<char>, bool> >' is different in different modules
    struct _Rb_tree_iterator
           ^
/usr/lib/gcc/x86_64-pc-linux-gnu/11.2.0/../../../../include/c++/11.2.0/bits/stl_tree.h:315:7: error: redefinition of 'operator=='
      operator==(const _Self& __x, const _Self& __y) _GLIBCXX_NOEXCEPT
      ^
/usr/lib/gcc/x86_64-pc-linux-gnu/11.2.0/../../../../include/c++/11.2.0/bits/stl_tree.h:1102:24: note: in instantiation of template class 'std::_Rb_tree_iterator<std::pair<const std::__cxx11::basic_string<char>, bool> >' requested here
            _M_insert_unique_(end(), *__first, __an);
                              ^
/usr/lib/gcc/x86_64-pc-linux-gnu/11.2.0/../../../../include/c++/11.2.0/bits/stl_map.h:232:14: note: in instantiation of function template specialization 'std::_Rb_tree<std::__cxx11::basic_string<char>, std::pair<const std::__cxx11::basic_string<char>, bool>, std::_Select1st<std::pair<const std::__cxx11::basic_string<char>, bool> >, std::less<std::__cxx11::basic_string<char> >, std::allocator<std::pair<const std::__cxx11::basic_string<char>, bool> > >::_M_insert_range_unique<const std::pair<const std::__cxx11::basic_string<char>, bool> *>' requested here
      { _M_t._M_insert_range_unique(__l.begin(), __l.end()); }
             ^
/home/blue/Scratchpad/work/forum_rvec_crash_2/repro_modules_crash.C:2:31: note: in instantiation of member function 'std::map<std::__cxx11::basic_string<char>, bool, std::less<std::__cxx11::basic_string<char> >, std::allocator<std::pair<const std::__cxx11::basic_string<char>, bool> > >::map' requested here
  std::map<std::string, bool> m{{"foo", true}, {"bar", false}};
                              ^
/usr/lib/gcc/x86_64-pc-linux-gnu/11.2.0/../../../../include/c++/11.2.0/bits/stl_tree.h:315:7: note: previous definition is here
      operator==(const _Self& __x, const _Self& __y) _GLIBCXX_NOEXCEPT
      ^
```

Note 1: A `Release` build seems to work fine.

I can reproduce the problem on CentOS 8 Stream with a self-built GCC 11.1.0. When testing with self-built versions of GCC 11.2.0, 10.2.0, and 9.3.0, I get a probably related crash `ROOT/src/interpreter/llvm/src/tools/clang/lib/AST/Decl.cpp:3857: unsigned int clang::FunctionDecl::getODRHash() const: Assertion 'hasODRHash()' failed.` where the `FunctionDecl` is the constructor of `std::_Rb_tree_iterator`.

Note 2: The system-default GCC 8.5.0 and a self-built GCC 8.4.0 are both fine. The only "relevant" change that I see for `_Rb_tree_iterator` in GCC 9 is the move from (member) `bool operator==` to `friend bool operator==`, which matches the error message quite nicely.

Finally, I was able to reproduce the assertion in a Docker container of `ubuntu:latest` (Ubuntu 22.04.4 LTS at the time of writing), after installing
```
 # apt update
 # apt upgrade
 # apt install --no-install-recommends ca-certificates cmake gcc g++ git libc6-dev libz-dev make ninja-build patch python3 uuid-dev
```
and then configuring as `cmake -G Ninja -DCMAKE_BUILD_TYPE=Debug -Dx11=OFF ../root.src/`.

**Comment by Axel-Naumann** [comment]
FYI @vgvassilev - module merge issue? Shouldn't `bits/stl_tree.h` be part of a module?

**Comment by vgvassilev** [comment]
We have discussed that issue. It is not trivial to make `bits/stl_tree.h` a part of a single submodule due to some extra entanglement (in past? stl versions). The expectation is this to be fixed with the upcoming upgrade. Is it still an issue there?

**Comment by hahnjo** [comment]
Yes, I can reproduce this issue with the upgrade to LLVM 13

**Comment by vgvassilev** [comment]
Not sure if that'd help but can you try with this version of the modulemap https://github.com/root-project/root/pull/11874/commits/b4af76caef9ea4504f99486f047d88b03b021d2d available in PR #11874?

**Comment by hahnjo** [comment]
So, the bad news is that I now get a crash on Arch (`Debug` build with GCC 12.2.1), also with https://github.com/root-project/root/pull/12062, and the other version of the modulemap doesn't seem to help...

**Comment by hahnjo** [comment]
I cannot reproduce this anymore with current `master`, neither on Arch (which upgraded to GCC 13 in the meantime) nor on EL8 with self-built GCCs. I attempted to bisect which change fixed it using a Debug build with GCC 12.2.0 and that points to commit 77127ba6537e54c91141bb11373d9a460f6c813f which is hard to believe (but I verified twice)...

cc @eguiraud

**Comment by eguiraud** [comment]
Hello, I must admit I don't even remember reporting this :sweat_smile: but the original repro works fine for me on Arch and gcc 13.

If the commit that "fixes" this is completely unrelated it might be that it is just shuffling enough bytes around the right way to hide the original problem. It's a pity the cause was never understood.

**Comment by hahnjo** [comment]
> Hello, I must admit I don't even remember reporting this :sweat_smile:

If memory serves well, it was somewhere on Mattermost - "this doesn't work in a Debug build, maybe somebody familiar with Cling wants to take a look" :smiley: 

> If the commit that "fixes" this is completely unrelated it might be that it is just shuffling enough bytes around the right way to hide the original problem. It's a pity the cause was never understood.

Yes, that's what I didn't like as well and why I started looking for the commit that fixed it. The reason I came back to this now is that I suspected addressing this via https://github.com/root-project/root/pull/13046 which would match well my memory about debugging this one a few months ago and declaring it related to our problems on macOS. And true enough, I've now applied commit ffdc579bbcd858e966db879c4258d0623eb520f1 on top of a6bc232098b61fa5c292de720e19de9ed457e77c (the parent of 77127ba6537e54c91141bb11373d9a460f6c813f mentioned above) - an idea I didn't have yesterday - and I can confirm it fixes the original issue in a version that reproduced the problem. So I'm going to mark this as fixed :smile:
